### PR TITLE
PR for katafourteen.py exercise

### PR DESCRIPTION
**Pull request by LouReis**
Please review katafourteen.py.  This was a confusing assignment but I did create a program that takes text files such as sherlock or sherlock_small as well as others, then processes the file creating a dictionary.  The dictionary uses two word keys while adding the third word as a value for that key.  There weren't any clear requirements on punctuation, proper sentences, upper/lower case, but the result will not have proper sentence structure making any punctuation / capitalization a problem.  My program creates a random 300 word short story derived from the dictionary.

Thanks & Regards,
Lou

### CODE REVIEW

**Comment by LouReis** [comment]
Hello, This PR is also for Lesson05 Exceptions Exercise.  Please review except_exercise.py as well.

Thanks & Regards,
Lou

**Comment by LouReis** [comment]
Hello, This PR is also for Lesson05 Mailroom3.py.  Please review mailroom3.py as well.  I read through the Lesson05 Comprehensions reading material a second time but I did not see a nifty way to use comprehensions in mailroom3.py.  My loops / assignments are involved and do several things and implementing comprehensions didn't seem to fit anywhere.  

Thanks & Regards,
Lou

File: students/LouReis/Lesson04/katafourteen.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3.7
+# katafourteen.py
+# Coded by LouReis
+
+# Take a text file as input and for each set of 3 words, make the first two words keys.
+
+print('This program takes a file and creates a dictionary.\n')
+print('Here are the files you have to choose from in this directory:\n\n')
+import pathlib
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson04/katafourteen.py]: all imports should go to the top of the file

File: students/LouReis/Lesson04/katafourteen.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+I wish I may I wish I might
+
+You might generate:
+
+"I wish" => ["I", "I"]
+"wish I" => ["may", "might"]
+"may I"  => ["wish"]
+"I may"  => ["I"]
+
+"""
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson04/katafourteen.py]: I fed in a file with this text "I wish I may I wish I might" and I don't seem to be getting the right mappings. I always get this:
I wish => ['I']
may I => ['wish']

----
from "You might generate:" ... your dictionary should have those keys, but values may be a bit different. You're however, only generating two keys.

Please review my sample solution here to see if it makes more sense:
https://github.com/natasha-aleksandrova/Self_Paced-Online/blob/master/solutions/lesson4/trigrams.py

File: students/LouReis/Lesson05/mailroom3.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+def main_menu(main_prompt,menu_options_dict):
+    while True:
+        try:
+            response = input(main_prompt)
+            menu_options_dict[response]()
+        except KeyError:
+            print("\n\n----------------PLEASE TRY AGAIN! PLEASE ENTER A VALID VALUE!----------------\n\n")
+            print("\n\n----------------PLEASE TRY AGAIN! PLEASE ENTER A VALID VALUE!----------------\n\n")
+            print("\n\n----------------PLEASE TRY AGAIN! PLEASE ENTER A VALID VALUE!----------------\n\n")
+            main_menu(main_prompt,menu_options_dict)
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson05/mailroom3.py]: no need for this line, your loop will continue looping - here you're creating another nested while True loop

File: students/LouReis/Lesson05/mailroom3.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+# The following function takes the 'donations' dictionary and creates a sorted report.
+# Using a complex 'sorted' for loop.
+def donation_report():
+    print('\nYou Chose Option 1\n\n')
+    print('DONATION SUMMARY REPORT\n\n')
+    print('{:25} | {:^13} | {:^13} |   {:>13}'.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift'))
+    print('---------------------------------------------------------------------------')
+    print()
+    for key,value in sorted(donations.items(),key=lambda i:sum(i[1]),reverse=True):
+        total = 0
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson05/mailroom3.py]: no need for this, you're redefining same var below.

File: students/LouReis/Lesson05/mailroom3.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        total = sum(donations[key])
+        count = 0
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson05/mailroom3.py]: no need for this

File: students/LouReis/Lesson05/mailroom3.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            main_menu(main_prompt,menu_options_dict)
+        donations[donor] = donations[donor] + [donation]
+    else:
+        print("You have entered a new donor:", donor)
+        donation = []
+        try:
+            donation = float(input("Please enter the donation amount '0.00':"))
+        except ValueError:
+            print("\n\n----------You have entered an invalid value, returning to Main Menu----------\n\n")
+            main_menu(main_prompt,menu_options_dict)
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson05/mailroom3.py]: exit the function, don't call the main menu again.

File: students/LouReis/Lesson05/mailroom3.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        print("You have entered an existing donor:", donor)
+        try:
+            donation = float(input("Please enter the donation amount '0.00':"))
+        except ValueError:
+            print("\n\n----------You have entered an invalid value, returning to Main Menu----------\n\n")
+            main_menu(main_prompt,menu_options_dict)
+        donations[donor] = donations[donor] + [donation]
+    else:
+        print("You have entered a new donor:", donor)
+        donation = []
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson05/mailroom3.py]: why is this defined as a list? no need for this as you're redefining it below

File: students/LouReis/Lesson05/mailroom3.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        filename = ""
+        filename = item.replace(" ","") + '.txt'
+        print(filename)
+        with open(filename, 'w') as f:
+            for i in range(1):
+                f.write(f"\n\nSubject: Donation\n\nDear {item},\n\nThank you for your donation, it will be used to help meet our goals.")
+                f.write("\nWe will welcome any future donations and appreciate your support.")
+                f.write("\n\n\nSincerely,\n\nMDTS Staff\n\n\n")
+    print("\n\nA Letter has been created for each donor and stored in a text file.\n\n")
+    return "Menu"
```
Comment (natasha-aleksandrova) [inline students/LouReis/Lesson05/mailroom3.py]: is this return needed?

**Pull request merged**